File: app/src/main/java/hu/epam/worktime/mvvmworkdroid/common/widgets/recyclerview/RecyclerViewAdapter.java
package hu.epam.worktime.mvvmworkdroid.common.widgets.recyclerview;

import java.util.Collections;
import java.util.List;

import com.hcom.android.BR;

import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

/**
 * Common adapter which is used for each recyclerview when using databinding.
 * Created by Mate_Redecsi on 10/20/2016.
 */
public class RecyclerViewAdapter extends RecyclerView.Adapter<BindingViewHolder> {
    private int column;
    private boolean displayFullRows;
    private List<ListItemViewModel> items;
    private List<Integer> viewTypes;
    private List<Integer> itemLayouts;

    public RecyclerViewAdapter() {
    }

    @Override
    public BindingViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int itemLayout = itemLayouts.get(viewTypes != null ? viewTypes.indexOf(viewType) : viewType);
        ViewDataBinding dataBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), itemLayout, parent, false);
        return new BindingViewHolder(dataBinding);
    }

    @Override
    public void onBindViewHolder(BindingViewHolder holder, int position) {
        holder.getBinding().setVariable(BR.viewModel, items.get(position));
        holder.getBinding().executePendingBindings();
    }

    @Override
    public int getItemCount() {
        int itemCount = items == null ? 0 : items.size();
        if (itemCount > column && displayFullRows) {
            itemCount = getRoundDownItemCount(itemCount);
        }
        return itemCount;
    }

    @Override
    public int getItemViewType(int position) {
        return items.get(position).getViewType();
    }

    public void setItems(List<ListItemViewModel> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    public void setItemLayout(int itemLayout) {
        this.itemLayouts = Collections.singletonList(itemLayout);
    }

    public void setItemLayouts(List<Integer> itemLayouts) {
        this.itemLayouts = itemLayouts;
    }

    public void setViewTypes(List<Integer> viewTypes) {
        this.viewTypes = viewTypes;
    }

    public void setColumn(int column) {
        this.column = column;
    }

    public void displayFullRows(boolean displayFullRows) {
        this.displayFullRows = displayFullRows;
    }

    private int getRoundDownItemCount(int itemCount) {
        return itemCount - itemCount % column;
    }
}

File: app/src/main/java/hu/epam/worktime/mvvmworkdroid/common/widgets/recyclerview/RecyclerViewAdapter.kt
package hu.epam.worktime.mvvmworkdroid.common.widgets.recyclerview

import java.util.Collections

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import hu.epam.worktime.mvvmworkdroid.BR

/**
 * Common adapter which is used for each recyclerview when using databinding.
 * Created by Mate_Redecsi on 10/20/2016.
 */
class RecyclerViewAdapter : RecyclerView.Adapter<BindingViewHolder>() {
    private var column: Int = 0
    private var displayFullRows: Boolean = false
    private var items: List<ListItemViewModel>? = null
    private var viewTypes: List<Int>? = null
    private var itemLayouts: List<Int>? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingViewHolder {
        val itemLayout = itemLayouts!![if (viewTypes != null) viewTypes!!.indexOf(viewType) else viewType]
        val dataBinding = DataBindingUtil.inflate<ViewDataBinding>(LayoutInflater.from(parent.context), itemLayout, parent, false)
        return BindingViewHolder(dataBinding)
    }

    override fun onBindViewHolder(holder: BindingViewHolder, position: Int) {
        holder.binding.setVariable(BR.viewModel, items!![position])
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        var itemCount = if (items == null) 0 else items!!.size
        if (itemCount > column && displayFullRows) {
            itemCount = getRoundDownItemCount(itemCount)
        }
        return itemCount
    }

    override fun getItemViewType(position: Int): Int {
        return items!![position].viewType
    }

    fun setItems(items: List<ListItemViewModel>) {
        this.items = items
        notifyDataSetChanged()
    }

    fun setItemLayout(itemLayout: Int) {
        this.itemLayouts = listOf(itemLayout)
    }

    fun setItemLayouts(itemLayouts: List<Int>) {
        this.itemLayouts = itemLayouts
    }

    fun setViewTypes(viewTypes: List<Int>) {
        this.viewTypes = viewTypes
    }

    fun setColumn(column: Int) {
        this.column = column
    }

    fun displayFullRows(displayFullRows: Boolean) {
        this.displayFullRows = displayFullRows
    }

    private fun getRoundDownItemCount(itemCount: Int): Int {
        return itemCount - itemCount % column
    }
}
